File: Problemset/remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java

// @Title: 删除链表的倒数第N个节点 (Remove Nth Node From End of List)
// @Author: 15218859676
// @Date: 2020-08-16 11:01:52
// @Runtime: 0 ms
// @Memory: 38.1 MB

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // int i = 0;
        // ListNode  pre = new ListNode(0);
        // pre.next = head;
        // while(i < n ){
        //     head = head.next;
        //     i++;
        // }
        // ListNode cur = pre.next;
        // if(head == null){
        //     pre.next = pre.next.next;
        //     return pre.next;
        // }
        // while (head.next!=null){
        //     cur = cur.next;
        //     head = head.next;
        // }
        // cur.next = cur.next.next;
        // return pre.next;
        ListNode pre = new ListNode(0);
        pre.next = head;
        ListNode start = pre, end = pre;
        while(n != 0) {
            start = start.next;
            n--;
        }
        while(start.next != null) {
            start = start.next;
            end = end.next;
        }
        end.next = end.next.next;
        return pre.next;


    }
}

File: Problemset/remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py

# @Title: 删除链表的倒数第N个节点 (Remove Nth Node From End of List)
# @Author: 15218859676
# @Date: 2020-08-16 14:52:13
# @Runtime: 24 ms
# @Memory: 12.7 MB

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        # pre = ListNode(0)
        # pre.next = head
        
        # i = 0 
        # while i < n :
        #     head  = head.next
        #     i+=1
        # if not head :
        # # if head is None:
        #     pre.next = pre.next.next;
        #     return pre.next
        # cur = pre.next
        # while head.next :
        # # while head.next is not None:
        #     cur = cur.next
        #     head = head.next
        # cur.next = cur.next.next
        # return pre.next
        pre = ListNode(0)
        pre.next = head
        fast = pre
        slow = pre
        i = 0 
        while i < n :
            fast = fast.next
            i += 1
        while fast.next:
            fast = fast.next
            slow = slow.next
        slow.next = slow.next.next
        return pre.next




